Represent non-fixed params as `NaN` for `QkTargetOp`

raynelfss opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Fixes #15438 
Due to our decision of provisionally exposing parameters in `QkTargetOp` as `double[]` we mistakenly filtered out any non-fixed parameters instead of showing a placeholder value. This PR provisionally exposes any non-fixed parameter instance as `NaN` until we work out a much cleaner way of exposing the parameters to C.

### Details and comments
- [x] Add release note.

raynelfss committed:
Initial: Represent non-fixed params as `NaN`

Cryoris commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -1343,8 +1343,9 @@ pub struct CTargetOp {
-    /// of `double`. If the operation doesn't posess any fixed parameters
-    /// or is variadic, this attribute will be a ``NULL`` pointer.
+    /// of `double`. If any of the parameters represented are not fixed angles
+    /// it will be represented as with the `NaN` value. If there are no parameters
+    /// then this value will be represented with a `NULL` pointer.
```
This last part isn't true anymore, is it? If all parameters are free then we'd still have an array with all `f64::NAN` entries, right?

raynelfss replied:
I guess I could've worded it better. But what I meant was if the operation doesn't accept any parameters.

jakelishman replied:
I think this comment is correct for the implementation now.

Cryoris reviewed (commented):
From a bugfix POV I don't think we need to backport this (since we can't create a C Target object that would have `params` of unknown length). But from an interface POV this PR changes an all-free `params` from a NULL pointer to an array of `f64::NAN` and it would be good to not change this interface between 2.3 and 2.4, which makes it backportable IMO.

eliarbel commented on releasenotes/notes/c-api-target-params-848091afd753a6c3.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+upgrade_c:
```
Do we still want this for 2.3? if so, this shouldn't be an upgrade but rather a feature which is actually already covered in #15512, so for 2.3 we actually don't need the reno at all.

raynelfss replied:
Sounds good, I will remove the reno.

Edit: Removed in 1568decd2d25ac6b95bfb0ff5fd592a3934f919d

eliarbel commented on test/c/test_target.c:
```diff
@@ -1022,6 +1022,24 @@ static int test_target_operation(void) {
                 result = EqualityError;
                 goto cleanup;
             }
+            if (inst_idx == 1) {
```
This comment is for the entire for-loop in this function - we need to call `qk_target_op_clear(&op)` at the end of each iteration otherwise the slices allocated in `qk_target_op_get` are not being freed.

raynelfss replied:
Added in 3b5d31324d9ab0fec13529885b6b9f3156112f28

eliarbel reviewed (commented):
Overall, this LTGM and almost ready to merge, thanks Ray! I had a couple minor comments, please see them inline. 

@Cryoris do you have any other comments?

eliarbel reviewed (dismissed):
LGTM, thanks for the changes. I'll go ahead and approve but will not enqueue for now in case Julien has more comments.

Cryoris commented on test/c/test_target.c:
Do we have a test that checks the parameters are indeed `NULL` for free parameters?

jakelishman reviewed (changes requested):
imo changing the typing like this is too big a change to merge after rc1 is released and the same day as the release.  I would have been comfortable leaving all the typing in place and using `nan` as a sentinel for "free parameter" for backport within the rc period, but swapping to `QkParam` has additional borrowing semantics that we haven't checked.

The C API is explicitly unstable and can be changed between versions, while the rc period is a well-established development practice we have to stop us from rushing in major untested changes even to _stable_ APIs.  I'd much rather we stick to the spirit of the rc freeze than rush in changes to an API that we can change in the next release.

raynelfss committed:
Docs: Add release note

raynelfss commented:
I reverted the PR to its original state for an easier patch. Added a fix release note and kept the commits adding ``QkParam`` in a different branch, for which I will open a new PR.

jakelishman commented on releasenotes/notes/c-api-target-params-6a54fcfcbe14de7e.yaml:
```diff
@@ -0,0 +1,8 @@
+    stored in :c:member:`QkTargetOp.params` did not match the number exposed by
+    :c:member:`QkTargetOp.num_params`. The following PR fixes it by exposing any
+    placeholder parameters as ``NAN`` so the length of the array always matches
+    that of :c:member:`QkTargetOp.num_params`.
```
We can't say "the following PR" since the release notes don't refer to PRs.  We might also want to comment that the bug is only present in rc1 (I think?).

Maybe something like

> Wildcard parameters are now represented by `nan`, and the length of the array will always be `` :c:member:`QkTargetOp.num_params` ``.  This bug was only present in 2.3.0rc1.

?

raynelfss replied:
Changed in 22abd7d3518ff018ef677ed2875cff7614c28b9e

jakelishman reviewed (commented):
Thanks Ray, this looks much more suitable for backport.

raynelfss committed:
Docs: Removed incorrect wording in release note.

Co-authored-by: Jake Lishman <jake.lishman@ibm.com>

jakelishman reviewed (approved):
Thanks Ray - I think we're good to go now.
